### ISSUE TITLE
pyre 0.0.23 and 0.0.24 crashing with SharedMemory exception

### ISSUE THREAD

**Issue by dseevr** [facebook/pyre-check#158]
`0.0.22` is _not_ crashing, however.  This is on an Ubuntu 18.04 (x64) system using Python 3.7.3

Here's the output I see on a `--debug` run of `0.0.24`:

```
2019-04-29 09:51:04,729 WARNING Could not parse 22 external files due to syntax errors!
2019-04-29 09:51:04,829 PERFORMANCE Stubs parsed: 8.907000s
2019-04-29 09:51:40,474 INFO Parsing 136 sources in `/redacted`...
2019-04-29 09:51:40,577 PERFORMANCE Sources parsed: 0.142000s
2019-04-29 09:51:40,578 PERFORMANCE Registered ignores: 0.063000s
2019-04-29 09:51:40,580 INFO Adding environment information to shared memory...
2019-04-29 09:51:40,580 PERFORMANCE Added environment to shared memory: 0.001000s
2019-04-29 09:51:40,681 INFO Building type environment...
2019-04-29 09:53:57,634 ERROR Client exited with error code 1:
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
Worker interrupted with signal: sigpipe
```

Let me know if I can provide any more info.  I've reverted to `0.0.22` to continue using pyre in the meanwhile.

**Comment by sinancepel** [comment]
@dseevr Are you able to replicate this crash with an empty source directory? I'm trying to see whether the issue has to do with your stubs/external sources. (An easy way of achieving this would be taking the same configuration, and pointing the source directory to be a newly created directory that has a dummy file).

**Comment by dseevr** [comment]
I set `--source-directory` to an empty dir and it worked fine. 🤔 

I am setting two `--exclude`s, two `--search-path`s, and `--source-directory`.  For this GH issue, I am also setting `--debug`.

**Comment by grievejia** [comment]
Well then it seems that the error may come from files in your project. I looked through the commits between 0.0.22 and 0.0.23 and did not have success in finding a suspicious diff. It might not be realistic to ask you to share the entire project source with us. But it would be nice if we can get a small repro would be really helpful for triaging the issue.

**Comment by dseevr** [comment]
@grievejia Thanks for looking, I'll see if i can narrow it down to a minimal reproducible case

**Comment by JelleZijlstra** [comment]
I got this crash too running pyre on my codebase:
```
2019-05-07 12:45:31,243 ERROR Client exited with error code 1:
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
```
This may have been because I passed the same directory in `--search-path` and `--source-directory`.

**Comment by mlin** [comment]
+1 -- our `.pyre_configuration` provides both `source_directories` and `search_path`, and getting this `Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")` with the recent versions.

**Comment by dkgi** [comment]
I have a repro for this internally. Will update once we figure out what's going on.

**Comment by grievejia** [comment]
I think I've found the cause of the issue. A fix is incoming but if you want to unblock yourself before that, the workaround is to avoid overlapping search path/source directory.

**Comment by dseevr** [comment]
I think this is a common use case in a Django application... your code is in `.`, your working directory is also `.`, and you import modules from one subdirectory into another using the Django application's root (`.` again) as a base rather than using relative imports.  (not a Django expert but this is what I've seen)

**Comment by grievejia** [comment]
Fixed in b77b775

**Comment by dseevr** [comment]
Thanks a lot for looking into this.  Unfortunately, I am still seeing a crash with 0.0.26:

```
# pyre --version
No watchman binary found.
To enable pyre incremental, you can install watchman: https://facebook.github.io/watchman/docs/install.html
Defaulting to non-incremental check.
 ƛ Found: `/home/redacted/.pyenv/versions/3.7.3/lib/pyre_check/typeshed/`
binary version: unversioned
client version: 0.0.26
```

<details>
<summary><strong>Click Here for the log output</strong></summary>

```
2019-05-17 16:00:43,525 INFO Found: `/home/redacted/.virtualenvs/redacted-NbpVAuKI/lib/pyre_check/typeshed/`
2019-05-17 16:00:43,525 WARNING Setting up a `.pyre_configuration` with `pyre init` may reduce overhead.
2019-05-17 16:00:43,525 DEBUG Running `/home/redacted/.virtualenvs/redacted/bin/pyre.bin check -debug -logging-sections parser -project-root /home/redacted -filter-directories /home/redacted -workers 36 -typeshed /home/redacted/.virtualenvs/redacted-NbpVAuKI/lib/pyre_check/typeshed/ -search-path /home/redacted -exclude /home/redacted -exclude /home/redacted /home/redacted`
2019-05-17 16:00:43,580 INFO Finding type stubs in `/home/redacted`...
2019-05-17 16:00:43,681 INFO Finding type stubs in `/home/redacted`...
2019-05-17 16:00:43,723 INFO Finding type stubs in `/home/redacted`...
2019-05-17 16:00:43,730 INFO Finding type stubs in `/home/redacted/.virtualenvs/redacted-NbpVAuKI/lib/pyre_check/typeshed/third_party`...
2019-05-17 16:00:43,735 INFO Finding type stubs in `/home/redacted/.virtualenvs/redacted-NbpVAuKI/lib/pyre_check/typeshed/stdlib`...
2019-05-17 16:00:43,835 INFO Finding external sources in `/home/redacted/.virtualenvs/redacted-NbpVAuKI/lib/python3.7/site-packages`...
2019-05-17 16:00:43,915 INFO Finding external sources in `/home/redacted`...
2019-05-17 16:00:44,016 INFO Parsing 6901 stubs and external sources...
2019-05-17 16:00:46,094 PERFORMANCE Stubs parsed: 2.415000s
2019-05-17 16:00:56,556 INFO Parsing 137 sources in `/home/redacted`...
2019-05-17 16:00:56,674 PERFORMANCE Sources parsed: 0.139000s
2019-05-17 16:00:56,675 PERFORMANCE Registered ignores: 0.082000s
2019-05-17 16:00:56,675 INFO Adding environment information to shared memory...
2019-05-17 16:00:56,676 PERFORMANCE Added environment to shared memory: 0.001000s
2019-05-17 16:00:56,776 INFO Building type environment...
2019-05-17 16:01:20,845 ERROR Client exited with error code 1:
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
Exception: SharedMem.C_assertion_failure("hh_shared.c : 1072")
Potential backtrace:
Raised at file "map.ml", line 131, characters 10-25
Called from file "sexp_conv.ml", line 148, characters 10-37
Worker exited (code: 2)
```
</details>

Unfortunately, I have to redact a lot of the full paths.

I noticed that it was spawning 36 workers by default (I'm running pyre on a 40 core machine) so I tried re-running that subcommand with only 4 workers and it still crashed (just with fewer exceptions).

**Comment by grievejia** [comment]
I don't think we released a version that contains the fix. Reopen until next release.

**Comment by grievejia** [comment]
@dseevr I tried to cut a new 0.0.27 release. Doe it work for you?

**Comment by dseevr** [comment]
@grievejia yes, it's working great!  Execution time has also been halved:

0.0.22:

```
pyre --exclude  --exclude  --search-path  --search-path . --source-directory   20.39s user 2.39s system 86% cpu 26.366 total
```

0.0.27:

```
pyre --exclude  --exclude  --search-path  --search-path . --source-directory   10.34s user 1.10s system 84% cpu 13.591 total
```

Thank you for your hard work!

**Comment by mlin** [comment]
I tried upgrading to v0.0.27 and it doesn't throw this error, but also now doesn't discover or check any files in the `source_directories`. I wonder if the same thing happened to @dseevr causing it to "run" so much "faster."

```
2019-05-26 12:18:11,719 INFO Parsing 1030 stubs and external sources...
2019-05-26 12:18:12,383 PERFORMANCE Stubs parsed: 0.764000s
2019-05-26 12:18:12,383 INFO Parsing 0 sources in `/home/mlin/src/miniwdl/WDL`...
2019-05-26 12:18:12,391 PERFORMANCE Sources parsed: 0.000000s
```

cf. https://github.com/chanzuckerberg/miniwdl/blob/master/.pyre_configuration

I'd previously run into a problem where pyre was "succeeding" because it wasn't checking anything, so I introduced [this goofy regression test](https://github.com/chanzuckerberg/miniwdl/blob/79d874d96099f3d8848fc1eaec60d41c2a96ff0e/Makefile#L21) which temporarily writes out a source file with a type error and makes sure pyre catches it. That's how I noticed this right away...

**Comment by grievejia** [comment]
@mlin The way it works now is that `search_path` takes precedence over `source_directory` -- if a file is under one of the directory under `search_path`, it will be treated as external source and won't be type checked.

**Comment by mlin** [comment]
@grievejia thanks, that'd explain it...I wonder though was this an arbitrary choice or were you constrained to it? my intuition would be the other way around: if a file is found both as a source file and an external module, one should err on the side of typechecking it as a source file.

Above @dseevr had also described:

> I think this is a common use case in a Django application... your code is in ., your working directory is also ., and you import modules from one subdirectory into another using the Django application's root (. again) as a base rather than using relative imports. (not a Django expert but this is what I've seen)

I think the commonality is our search paths include the current directory `.`, our source is in a module subdirectory of the current directory (e.g. `./my_package/my_module.py`), and some other file uses an absolute import (`import my_package.my_module`). If we don't have `.` as a search path then the import would fail. 

I can probably switch to relative imports, but (i) there are [arguments for absolute imports](https://realpython.com/absolute-vs-relative-python-imports/) and (ii) the behavior that including `.` in your search path causes pyre to ~silently ignore everything seems like a clearly bad pitfall.

Thanks for the help! :beer:

**Comment by grievejia** [comment]
@mlin Thanks for the feedback, and sorry for the bad experience!

The reasoning behind the decision is that some users put the virtual environment under their source root. If `source_directory` overrides `search_path`, there's no way for them to exclude libraries installed in their virtualenv. 

The core issue here is that some users tend to put search path above source root, and some users like to put search path under source root. An arbitrating decision has to be made somehow. 

We do have a workaround  (albeit undocumented) if your use case falls into the first category: each element in `search_path` array can take an alternative form `{"root": "YOUR_SEARCH_ROOT", "subdirectory": "YOUR_SUBDIRECTORY"}`. The meaning is that pyre adds `YOUR_SUBDIRECTORY ` to search path, but absolute imports of the files under `YOUR_SUBDIRECTORY ` will be counted from `YOUR_SEARCH_ROOT`, if that makes sense. This can be used to circumvent the need of putting the source root (or any path above) into the search path.

**Comment by dseevr** [comment]
I guess the celebration _was_ a bit early.  😄 Good catch, @mlin.  I have confirmed that pyre 0.0.27 no longer catches type errors in my code with the same invocation that works properly with 0.0.22.

Since the original issue here (SharedMemory exception) was resolved, should I open another issue for figuring out how to properly use pyre with a Django project?  I'm going to stick with 0.0.22 in the meanwhile.

**Comment by mlin** [comment]
Thanks @grievejia, the virtualenv scenario helps me understand the reasoning. If you're keeping this behavior, I think it's worth at least displaying some warning that files under the source directories are being skipped because they're in the search path.

I don't think the workaround quite helps me unfortunately, as I've got absolute imports of sibling modules within the same package subdirectory. I'll need to come back to this when I have some more time and switch them to relative imports (and also stick with v0.0.22 in the meantime). I'm not as familiar with Django or whether the workaround helps @dseevr

**Comment by mlin** [comment]
Just pinging this

I've now made all of my package's internal imports relative, which allowed me to upgrade to v0.0.27. So I'm good with this, but I bet others are going to get burned by including their package directory in both `source_directory` and `search_path`.

**Comment by dseevr** [comment]
I'm still stuck on 0.0.22 because I can't figure out how to run the latest version against a Django project

**Comment by dseevr** [comment]
Should we keep this issue open?  The exception has been fixed, but now the tool itself no longer works correctly for me (as it did in the past) so I'm stuck on an old version.  That almost seems like a discussion that would warrant its own issue, though.

**Comment by grievejia** [comment]
@dseevr  @mlin Sorry for the long turnaround time! A new release has just been published, which contains commit 545d5eabff8ca9a2d5a1697b05358b4abb8d23dc that may address this issue. 

By default, everything under `source_directory` should now get type checked, unless they are explicitly blacklisted by a configuration option `ignore_all_errors`. So yeah, we basically turned things around for you guys 😆 

The full story of how Pyre searches for source files and how Pyre makes the decision of whether to check a file or not is, however, slightly more complicated than that. My teammates and I went through multiple rounds of discussion on this issue, and we believe we now settled on a story that should be reasonable to tell. Later I'll update the documentation on the website with details on how things currently work.
